### PR TITLE
Domain entities: distinguish between matching requirements by field name & type

### PR DESCRIPTION
**Pull request by sbelak**
This introduces 3 changes:
1) domain entities only expose dimensions specified in `required_attributes` or `optional_attributes`
2) adds conflict resolution logic if multiple fields match the same attribute specification
3) [potentially controversial] adds explicit support to bind domains by type (preferred) or name

While matching by name is icky, I think there are cases where it's unavoidable. The two main ones are:
1) outputs from transforms. We can't expect that resulting cols in trasforms will always uniquely map to out types, especially when we do any sort of aggregations (it's reasonable to have a `price` type, much less to cover all the aggregation permutations such as `avg price`, `min price`, ...), and we don't have a good mechanism of extending types on the fly (while you can just drop in new transforms).
2) it makes writing vendor-specific trasforms (ie. the outermost layer of transforms) much easier and more sane as again you don't have to first account for all their particularities in our type system. This one is even nastier than 1) as it would entail also writing classifiers.

### CODE REVIEW

File: test/metabase/api/automagic_dashboards_test.clj [hunk 1]
```diff
@@ -171,9 +171,9 @@
 ;;; ------------------- Transforms -------------------
 
 (expect
-  [[4 1 10.0646 -165.374 "Red Medicine" 3 1 4 3 2 1]
-   [11 2 34.0996 -118.329 "Stout Burgers & Beers" 2 2 11 2 1 1]
-   [11 3 34.0406 -118.428 "The Apple Pan" 2 2 11 2 1 1]]
+  [[3 4 -165.374 10.0646 1 "Red Medicine" 1 4 3 2 1]
+   [2 11 -118.329 34.0996 2 "Stout Burgers & Beers" 2 11 2 1 1]
+   [2 11 -118.428 34.0406 3 "The Apple Pan" 2 11 2 1 1]]
```
Comment (salsakran) [inline test/metabase/api/automagic_dashboards_test.clj]: why are these changing?

Comment (sbelak) [reply test/metabase/api/automagic_dashboards_test.clj]: Just the order changed due to how we construct the `:dimensions` object (low cardinality maps in Clojure retain insertion order).

**Review by camsaul — APPROVED** [review]
wfm

**Comment by rlotun** [comment]
closing due to age.

**Pull request closed without merging**